### ISSUE TITLE
Task :migrateToRefreshVersionsDependenciesConstants/refreshVersions FAILED

### ISSUE THREAD

**Issue by HLQ-Struggle** [jmfayard/refreshVersions#251]
Hi @LouisCAD ,I have been trying to use refreshVersions recently to optimize the dependencies used in the gradle files of my project.

I am sure that after checking the documentation many times, the current method is correct, but I don’t know why it still makes an error. I hope to get your advice.

My expected result:

> Still based on the buildSrc method, split all the dependencies used in my project, such as all the dependencies in app/weight/helper, into buildSrc, such as Libs or Versions, and the build.gradle in the actual module can be directly Point attributes using the type name. And it also supports detecting whether the current dependency has the latest version.

In the first step, I upgraded my gradle to 6.4.1

> ./gradlew wrapper --gradle-version 6.4.1

The second step is to modify the contents of settings.gradle in the root directory

```
// settings.gradle.kts
import de.fayard.refreshVersions.RefreshVersionsSetup

// Here you might have some pluginManagement block:
pluginManagement {
     //...
}

buildscript {
     repositories {gradlePluginPortal()}
     dependencies.classpath("de.fayard.refreshVersions:refreshVersions:0.9.5")
}

rootProject.name ='FingerAdvisor_Android'
include':app'
include':helper'
include':weight'

RefreshVersionsSetup.bootstrap(settings)
```

Then began to execute the command:

> ./gradlew migrateToRefreshVersionsDependenciesConstants --console=plain

I have been studying for a long time, and I don’t know where the problem occurred.

```
➜  xxx_Android git:(master) ✗ ./gradlew migrateToRefreshVersionsDependenciesConstants --console=plain

> Task :migrateToRefreshVersionsDependenciesConstants
All the following modules have hardcoded dependencies versions

1. :app (7)
2. :helper (1)
3. :weight (1)
4. Exit

Type the number of the Gradle module you want to migrate first:
weight

> Task :migrateToRefreshVersionsDependenciesConstants FAILED
> Task :refreshVersions FAILED

FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':migrateToRefreshVersionsDependenciesConstants'.
> java.util.InputMismatchException (no error message)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':refreshVersions'.
> Couldn't resolve version for ModuleId(group=androidx.test.ext, name=junit)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

* Get more help at https://help.gradle.org

BUILD FAILED in 10s
2 actionable tasks: 2 executed
➜  FingerAdvisor_Android git:(master) ✗ 

```

![image](https://user-images.githubusercontent.com/19400093/93183967-44be3200-f76e-11ea-82a0-1a5d9a6fd076.png)

**Comment by LouisCAD** [comment]
Read the instructions in the instructions.
> Type the number of the Gradle module you want to migrate first:

**Comment by LouisCAD** [comment]
It asks for a **NUMBER** and you give it some text.

Just read the instructions, @HLQ-Struggle, it doesn't have to be a struggle.

**Comment by HLQ-Struggle** [comment]
Hi, @LouisCAD After your prompt, I adjusted the operation as follows:

![image](https://user-images.githubusercontent.com/19400093/93189540-d5980c00-f774-11ea-8194-a296e3965713.png)

After finishing, see the following prompt in the console:

```
> Task :refreshVersions
Found 50 hardcoded dependencies versions.

[com.android.tools.build:gradle:4.0.1, org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.71, com.android.tools.build:aapt2:4.0.1-6197926]...

To ensure single source of truth, refreshVersions only works with version placeholders,
that is the explicit way of marking the version is not there (but in the versions.properties file).

If you intentionally want to keep hardcoded versions so a module has a different version of a
dependency than the rest of the project, you can safely ignore this warning for these artifacts,
but keep in mind refreshVersions will not show available updates for these.

Note that a migration task is planned in a future version of refreshVersions.

See https://github.com/jmfayard/refreshVersions/issues/160

BUILD SUCCESSFUL in 17m 18s
2 actionable tasks: 2 executed
```

Regarding the problem you mentioned, I really don’t understand how to solve it?

![image](https://user-images.githubusercontent.com/19400093/93190178-9fa75780-f775-11ea-9214-25c0ddc7fb03.png)

Is it a placeholder like the following?

```
dependencies {
     implementation fileTree(dir:'libs', include: ['*.jar'])
     implementation Kotlin.stdlib.jdk7
     implementation AndroidX.appCompat
     implementation AndroidX.core.ktx

     testImplementation Testing.junit4
     androidTestImplementation AndroidX.test.ext.junit
     androidTestImplementation AndroidX.test.espresso.core

     api Google.android.material
     api "com.tencent:mmkv-static:_"
     api "com.airbnb.android:lottie:_"
}
```

**Comment by HLQ-Struggle** [comment]
@LouisCAD Thank you very much for taking the time to guide me as a rookie.

I would like to consult, because my original intention was to upgrade based on buildSrc, so I found this solution for you, but currently I do deal with it in the way of buildSrc. Why didn't I find the corresponding buildSrc directory and the corresponding files?

**Comment by LouisCAD** [comment]
Yes you are using the version placeholder correctly. BTW, instead of screenshotting issue comments, please give the link instead (get the permalink by clicking on the comment time/date next to the name of the commenter).

Great that you succeeded in using it correctly!

Regarding `buildSrc` we are providing an alternative to avoid you making frequent edits to `buildSrc` that would slow your builds down because of cache invalidations.

**Comment by HLQ-Struggle** [comment]
@LouisCAD Thank you so much for your guidance.

I want to know what is the alternative method you are talking about?

Because my original intention is that my Android project will contain multiple modules, each of which has its own dependencies. Relatively speaking, the buildSrc method combined with your upgrade reminder, I think this will be a perfect solution.

Regarding the speed of construction, there are pros and cons, and it is still acceptable for the time being. 😂

**Comment by HLQ-Struggle** [comment]
Hi @LouisCAD , The link you just proposed is as follows:

- [refreshVersions expect dependencies with a version placeholder _ #160 ](https://github.com/jmfayard/refreshVersions/issues/160)

I don't know why, but the issue will still be output when used correctly.

**Comment by HLQ-Struggle** [comment]
@LouisCAD My original intention was to add a detection version upgrade to the following example.

### buildSrc practice

Describe the operation steps:

- Create a buildSrc directory under the project root directory, and then create a new build.gradle.kts file;
- Create the src directory and the corresponding management version file;
- Replace the original dependencies directly

**build.gradle.kts content is as follows:**

```
// Import the Kotlin plugin
import org.gradle.kotlin.dsl.`kotlin-dsl`

plugins {
    `kotlin-dsl`
}

repositories {
    jcenter()
}

/**
 * Disable test reports (Gradle will automatically create test reports by default)
 */
tasks.withType<Test> {
    reports.html.isEnabled = false
    reports.junitXml.isEnabled = false
}

/**
 * isFork: Run the compiler as a separate process.
 * This process will be reused during construction, so the fork overhead is small. The benefit of forking is that memory-intensive compilation is performed in a different process, which results in a greatly reduced amount of garbage collection in the main Gradle daemon.
 * Less garbage collection in the daemon means that Gradle's infrastructure can run faster, especially if you still use --parallel.
 *
 * isIncremental: incremental compilation. Gradle can analyze dependencies down to the individual class level to recompile only the classes affected by the change. Since Gradle 4.10, incremental compilation is the default setting.
 */
tasks.withType<JavaCompile> {
    options.isFork = true
    options.isIncremental = true
}

/**
 * Disable warnings about using experimental Kotlin compiler features
 */
kotlinDslPluginOptions {
    experimentalWarning.set(false)
}
```

Dependencies.kt, this is the version management file that I defined. Part of the content is as follows:

```
@file:Suppress("SpellCheckingInspection")

/**
 * @author HLQ_Struggle
 * @date 2020/7/27
 * @desc unified management class
 */

// Unified management of version information in the project
object Versions {

    // Build Config
    const val compileSDK = 29 // compile SDK version
    const val buildTools = "29.0.3" // Gradle build tool version
    const val minSDK = 23 // Minimum compatible Android version
    const val targetSDK = 29 // The highest compatible Android version

    // App Version
    const val appVersionCode = 1 // current version number
    const val appVersionName = "1.0" // current version information

    // Plugins
    const val androidGradlePlugin = "4.0.1"

    // Kotlin
    const val kotlin = "1.3.72"
    const val kotlinxCoroutines = "1.3.5"

    // Support Lib
    const val support = "29.0.3"
    const val appcompat = "1.1.0"
    const val constrainLayout = "1.1.3"

    // Testing
    const val junit = "4.12"
    const val extJunit = "1.1.1"
    const val espresso = "3.2.0"
}

// Unified management of dependent libraries used in the project
object Deps {

    // Gradle
    const val androidGradle = "com.android.tools.build:gradle:${Versions.androidGradlePlugin}"

    // Kotlin
    const val kotlinStdLib = "org.jetbrains.kotlin:kotlin-stdlib-jdk7:${Versions.kotlin}"
    const val kotlinGradlePlugin = "org.jetbrains.kotlin:kotlin-gradle-plugin:${Versions.kotlin}"
    const val kotlinxCoroutines =
        "org.jetbrains.kotlinx:kotlinx-coroutines-android:${Versions.kotlinxCoroutines}"

    // Testing
    const val junit = "junit:junit:${Versions.junit}"
    const val extJunit = "androidx.test.ext:junit:${Versions.extJunit}"
    const val espresso = "androidx.test.espresso:espresso-core:${Versions.espresso}"

    // Android

    const val appcompat = "androidx.appcompat:appcompat:${Versions.appcompat}"
    const val coreKtx = "androidx.core:core-ktx:1.2.0"
    const val constraintLayout =
        "androidx.constraintlayout:constraintlayout:${Versions.constrainLayout}"

    // Fix: About 64K reference limit
    const val multiDex = "androidx.multidex:multidex:2.0.1"

    // Jetpack
    const val viewPager2 = "androidx.viewpager2:viewpager2:1.0.0"

// ...
}
```

Give two chestnuts, how to use:

- How to use build in the root directory:

You can get the corresponding attributes directly through the group name defined in the Dependencies file, as shown below:

```
buildscript {
    // ...
    dependencies {
        classpath Deps.androidGradle

        classpath Deps.kotlinGradlePlugin

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}
// ...
```

-How to use build under other module directory:

In the same way, of course, you can also directly import the entire corresponding grouping method and directly use the corresponding attributes, for example:

```
// The Deps and Versions grouping methods defined by directly importing are used here
import static Deps.*
import static Versions.*

apply plugin:'com.android.library'
apply plugin:'kotlin-android'
apply plugin:'kotlin-android-extensions'

android {

// Here you can use the corresponding attributes directly
    compileSdkVersion compileSDK
    buildToolsVersion buildTools

    defaultConfig {
        minSdkVersion minSDK
        targetSdkVersion targetSDK
        versionCode appVersionCode
        versionName appVersionName

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
        consumerProguardFiles'consumer-rules.pro'
    }
// ...
}

dependencies {
    implementation fileTree(dir:'libs', include: ['*.jar'])
    // The same is true here, just use the corresponding attribute name directly
    implementation kotlinStdLib
    implementation appcompat
    implementation coreKtx
    api'com.google.android.material:material:1.2.0'
    testImplementation junit
    androidTestImplementation extJunit
    androidTestImplementation espresso

    api mmkv
    api'com.airbnb.android:lottie:3.4.1'
}
```

This method has several good features as follows:

- Support jump;
- Support smart reminders;
- Gradle intervenes when compiling, feeling very wet and tall

But what about the key update prompt?

![image](https://user-images.githubusercontent.com/19400093/93194307-848b1680-f77a-11ea-9876-f16f9f26dbab.png)

**Comment by LouisCAD** [comment]
Please, stop pinging me for every single comment, please stop and sending extra long messages.

**Comment by LouisCAD** [comment]
> I don't know why, but the issue will still be output when used correctly.

It's a known issue. We hope to have it addressed before 1.0.

> I want to know what is the alternative method you are talking about?

This is what you're using: constants bundled into the refreshVersions' plugin.

> Regarding the speed of construction, there are pros and cons, and it is still acceptable for the time being. 😂

I'm not sure what you're referring to as "speed of construction".

**Comment by HLQ-Struggle** [comment]
Sorry for what I did just now.

The following example:

- Click to jump is not supported:

![image](https://user-images.githubusercontent.com/19400093/93197445-3a0b9900-f77e-11ea-9241-b98f286cb756.png)

- Although the replacement character is used, there is still a prompt that needs to be updated:

![image](https://user-images.githubusercontent.com/19400093/93197567-632c2980-f77e-11ea-8fe6-ba1ffdc40b6d.png)

It seems to go against my original intention:

- One-time citation, full-text (project) use, avoiding repeated dependencies introduced by teamwork;
- Comes with dependency update prompt;
- Support normal operations such as jumps (still not supported).

Is there any good solution?
